File: src/main/java/me/lukiiy/utils/cmd/DisplayToast.java
package me.lukiiy.utils.cmd;

import me.lukiiy.utils.cool.Presets;
import me.lukiiy.utils.cool.Toast;
import me.lukiiy.utils.main;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class DisplayToast implements CommandExecutor, TabExecutor {
    @Override
    public boolean onCommand(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] strings) {
        if (!(commandSender instanceof Player) && strings.length == 0) {
            commandSender.sendMessage(main.argsErrorMsg);
            return true;
        }
        Player target = strings.length > 0 ? Bukkit.getPlayer(strings[0]) : (Player) commandSender;
        if (target == null) {
            commandSender.sendMessage(main.notFoundMsg);
            return true;
        }

        Toast.Style style;
        try {style = Toast.Style.valueOf(strings[1].toUpperCase());}
        catch (Throwable ignored) {
            commandSender.sendMessage(Presets.Companion.warnMsg("Style not found!"));
            return true;
        }

        Material id;
        try {id = Material.valueOf(strings[2].toUpperCase());}
        catch (Throwable ignored) {
            commandSender.sendMessage(Presets.Companion.warnMsg("Material ID not found!"));
            return true;
        }

        String msg = String.join(" ", Arrays.copyOfRange(strings, 3, strings.length))
                .replace("/n", "\n")
                .replace("&", "§");
        boolean chatAnnounce = msg.contains("-c");
        if (chatAnnounce) msg = msg.replace("-c", "");

        commandSender.sendMessage(Presets.Companion.msg("Displayed custom Toast to ").append(target.name().color(Presets.Companion.getACCENT_NEUTRAL())));
        Toast.display(target, style, id, msg, chatAnnounce);
        return true;
    }

    @Override
    public @Nullable List<String> onTabComplete(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] strings) {
        List<String> tab = new ArrayList<>();
        switch (strings.length) {
            case 1 -> {return null;}
            case 2 -> {for (Toast.Style type : Toast.Style.values()) tab.add(type.name());}
            case 3 -> {for (Material id : Material.values()) tab.add(id.name().toLowerCase());}
        }
        return tab.stream().filter(next -> next.toLowerCase().startsWith(strings[strings.length - 1])).collect(Collectors.toList());
    }
}

File: src/main/java/me/lukiiy/utils/cmd/DisplayToast.kt
package me.lukiiy.utils.cmd

import com.mojang.brigadier.Command
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.tree.LiteralCommandNode
import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.Commands
import io.papermc.paper.command.brigadier.argument.ArgumentTypes
import me.lukiiy.utils.Defaults
import me.lukiiy.utils.help.Toast
import me.lukiiy.utils.help.Utils
import me.lukiiy.utils.help.Utils.asFancyString
import me.lukiiy.utils.help.Utils.asPermission
import me.lukiiy.utils.help.Utils.asPlainString
import me.lukiiy.utils.help.Utils.getPlayersOrThrow
import me.lukiiy.utils.help.Utils.mark
import me.lukiiy.utils.help.Utils.stripArgument
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.Style
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

object DisplayToast {
    private val main = Commands.literal("displaytoast")
        .requires { it.sender.hasPermission("displaytoast".asPermission()) }
        .then(Commands.argument("players", ArgumentTypes.players())
            .then(Commands.argument("style", StringArgumentType.word())
                .suggests { _, builder ->
                    val input = builder.remaining.uppercase()

                    Toast.Style.entries.forEach { if (it.name.lowercase().startsWith(input.lowercase())) builder.suggest(it.name.lowercase()) }
                    builder.buildFuture()
                }
                .then(Commands.argument("item", ArgumentTypes.itemStack())
                    .then(Commands.argument("message", StringArgumentType.greedyString())
                        .executes {
                            val sender = it.source.sender
                            val target = it.getPlayersOrThrow("players")

                            val style =
                                try { Toast.Style.valueOf(StringArgumentType.getString(it, "style").uppercase()) }
                                catch (_: IllegalArgumentException) { throw Defaults.CmdException(Component.text("Expected a valid style type.")) }

                            handle(sender, target, style, it.getArgument("item", ItemStack::class.java), StringArgumentType.getString(it, "message"))
                            Command.SINGLE_SUCCESS
                        }
                    )
                )
            )
        )

    fun register(): LiteralCommandNode<CommandSourceStack> = main.build()

    fun handle(sender: CommandSender, targets: List<Player>, style: Toast.Style, item: ItemStack, message: String) {
        val (msg, chatAnnounce) = message.replace("/n", "\n").replace("&", "§").stripArgument("c")
        val mark = targets.mark(Style.style(Defaults.YELLOW))

        Utils.adminCmdFeedback(sender, "Displyed a toast to ${mark.asPlainString()}")
        sender.sendMessage(Defaults.neutral("Displayed a toast to ".asFancyString().append(mark)))
        Toast.display(sender.name, targets, style, item, Defaults.FancyString.deserialize(msg), chatAnnounce)
    }
}